### ISSUE TITLE
cilium monitor prints nothing despite traffic passing traffic

### ISSUE THREAD

**Issue by danwent** [cilium/cilium#2641]
I'm not sure how reproducible this is, but my single node minikube is in the situation that it is passing traffic, but cilium monitor prints nothing.   Not only is the application still working, but the cilium access log is recording the traffic, so it seems clear that the traffic is passing through cilium.  

When i started out, cilium monitor was working, but then it stopped abruptly.  This may have been when I pushed down a kafka "allow-all" policy.   Even restarting cilium monitor and re-running the test traffic shows nothing.  

````
root@minikube:~# cilium monitor                                                                                                                                            
Listening for events on 2 CPUs with 64x4096 of shared memory
Press Ctrl-C to quit


-> forward (Reply connection), at To host, to endpoint 0, identity 289->1: 10.15.10.151:9092 -> 10.15.28.238:36806 tcp ACK
-> forward (Established connection), at To endpoint, to endpoint 22387, identity 292->289: 10.15.28.238:36806 -> 10.15.10.151:9092 tcp ACK
-> forward (Reply connection), at To endpoint, to endpoint 45959, identity 289->292: 10.0.0.134:9092 -> 10.15.161.37:36344 tcp ACK
-> forward (Established connection), at To proxy, to endpoint 0, identity 289->0: 10.15.161.37:36344 -> 10.15.10.151:9092 tcp ACK
-> forward (Established connection), at To proxy, to endpoint 0, identity 289->0: 10.15.161.37:36344 -> 10.15.10.151:9092 tcp ACK
-> forward (Established connection), at To endpoint, to endpoint 22387, identity 292->289: 10.15.28.238:36806 -> 10.15.10.151:9092 tcp ACK
-> forward (Established connection), at To endpoint, to endpoint 49574, identity 289->291: 10.15.10.151:52378 -> 10.15.173.191:2181 tcp ACK
-> forward (Reply connection), at To endpoint, to endpoint 22387, identity 291->289: 10.0.0.93:2181 -> 10.15.10.151:52378 tcp ACK
-> forward (Established connection), at To endpoint, to endpoint 49574, identity 289->291: 10.15.10.151:52378 -> 10.15.173.191:2181 tcp ACK
-> forward (Established connection), at To endpoint, to endpoint 49574, identity 289->291: 10.15.10.151:52378 -> 10.15.173.191:2181 tcp ACK
-> forward (Reply connection), at To endpoint, to endpoint 22387, identity 291->289: 10.0.0.93:2181 -> 10.15.10.151:52378 tcp ACK
-> forward (Established connection), at To endpoint, to endpoint 49574, identity 289->291: 10.15.10.151:52378 -> 10.15.173.191:2181 tcp ACK
-> forward (Reply connection), at To endpoint, to endpoint 22387, identity 291->289: 10.0.0.93:2181 -> 10.15.10.151:52378 tcp ACK
-> forward (Reply connection), at To host, to endpoint 0, identity 289->1: 10.15.10.151:9092 -> 10.15.28.238:36806 tcp ACK
-> forward (Reply connection), at To endpoint, to endpoint 45959, identity 289->292: 10.0.0.134:9092 -> 10.15.161.37:36344 tcp ACK
-> forward (Established connection), at To endpoint, to endpoint 49574, identity 289->291: 10.15.10.151:52378 -> 10.15.173.191:2181 tcp ACK
-> forward (Established connection), at To endpoint, to endpoint 22387, identity 292->289: 10.15.28.238:36806 -> 10.15.10.151:9092 tcp ACK
-> forward (Established connection), at To proxy, to endpoint 0, identity 289->0: 10.15.161.37:36344 -> 10.15.10.151:9092 tcp ACK
-> forward (Established connection), at To proxy, to endpoint 0, identity 289->0: 10.15.161.37:36344 -> 10.15.10.151:9092 tcp ACK
-> forward (Established connection), at To endpoint, to endpoint 22387, identity 292->289: 10.15.28.238:36806 -> 10.15.10.151:9092 tcp ACK
-> forward (Reply connection), at To host, to endpoint 0, identity 289->1: 10.15.10.151:9092 -> 10.15.28.238:36806 tcp ACK
-> forward (Established connection), at To endpoint, to endpoint 22387, identity 292->289: 10.15.28.238:36806 -> 10.15.10.151:9092 tcp ACK
-> forward (Reply connection), at To endpoint, to endpoint 45959, identity 289->292: 10.0.0.134:9092 -> 10.15.161.37:36344 tcp ACK
-> forward (Established connection), at To proxy, to endpoint 0, identity 289->0: 10.15.161.37:36344 -> 10.15.10.151:9092 tcp ACK
-> forward (Established connection), at To proxy, to endpoint 0, identity 289->0: 10.15.161.37:36344 -> 10.15.10.151:9092 tcp ACK
-> forward (Established connection), at To endpoint, to endpoint 22387, identity 292->289: 10.15.28.238:36806 -> 10.15.10.151:9092 tcp ACK
-> forward (Established connection), at To endpoint, to endpoint 49574, identity 289->291: 10.15.10.151:52378 -> 10.15.173.191:2181 tcp ACK
-> forward (Reply connection), at To endpoint, to endpoint 22387, identity 291->289: 10.0.0.93:2181 -> 10.15.10.151:52378 tcp ACK
-> forward (Established connection), at To endpoint, to endpoint 49574, identity 289->291: 10.15.10.151:52378 -> 10.15.173.191:2181 tcp ACK
-> forward (Established connection), at To endpoint, to endpoint 49574, identity 289->291: 10.15.10.151:52378 -> 10.15.173.191:2181 tcp ACK
-> forward (Reply connection), at To endpoint, to endpoint 22387, identity 291->289: 10.0.0.93:2181 -> 10.15.10.151:52378 tcp ACK
-> forward (Established connection), at To endpoint, to endpoint 49574, identity 289->291: 10.15.10.151:52378 -> 10.15.173.191:2181 tcp ACK
-> forward (Reply connection), at To endpoint, to endpoint 22387, identity 291->289: 10.0.0.93:2181 -> 10.15.10.151:52378 tcp ACK
-> forward (Reply connection), at To host, to endpoint 0, identity 289->1: 10.15.10.151:9092 -> 10.15.28.238:36806 tcp ACK
-> forward (Reply connection), at To endpoint, to endpoint 45959, identity 289->292: 10.0.0.134:9092 -> 10.15.161.37:36344 tcp ACK
-> forward (Established connection), at To endpoint, to endpoint 49574, identity 289->291: 10.15.10.151:52378 -> 10.15.173.191:2181 tcp ACK
-> forward (Established connection), at To endpoint, to endpoint 22387, identity 292->289: 10.15.28.238:36806 -> 10.15.10.151:9092 tcp ACK
-> forward (Established connection), at To proxy, to endpoint 0, identity 289->0: 10.15.161.37:36344 -> 10.15.10.151:9092 tcp ACK
-> forward (Established connection), at To proxy, to endpoint 0, identity 289->0: 10.15.161.37:36344 -> 10.15.10.151:9092 tcp ACK
-> forward (Reply connection), at To endpoint, to endpoint 45959, identity 289->292: 10.0.0.134:9092 -> 10.15.161.37:36344 tcp ACK
-> forward (Established connection), at To endpoint, to endpoint 22387, identity 292->289: 10.15.28.238:36806 -> 10.15.10.151:9092 tcp ACK
-> forward (Reply connection), at To host, to endpoint 0, identity 289->1: 10.15.10.151:9092 -> 10.15.28.238:36806 tcp ACK
-> forward (Established connection), at To endpoint, to endpoint 22387, identity 292->289: 10.15.28.238:36806 -> 10.15.10.151:9092 tcp ACK
-> forward (Reply connection), at To endpoint, to endpoint 45959, identity 289->292: 10.0.0.134:9092 -> 10.15.161.37:36344 tcp ACK
-> forward (Established connection), at To proxy, to endpoint 0, identity 289->0: 10.15.161.37:36344 -> 10.15.10.151:9092 tcp ACK
-> forward (Established connection), at To proxy, to endpoint 0, identity 289->0: 10.15.161.37:36344 -> 10.15.10.151:9092 tcp ACK
-> forward (Reply connection), at To endpoint, to endpoint 45959, identity 289->292: 10.0.0.134:9092 -> 10.15.161.37:36344 tcp ACK
-> forward (Established connection), at To endpoint, to endpoint 22387, identity 292->289: 10.15.28.238:36806 -> 10.15.10.151:9092 tcp ACK
-> forward (Established connection), at To endpoint, to endpoint 49574, identity 289->291: 10.15.10.151:52378 -> 10.15.173.191:2181 tcp ACK
-> forward (Reply connection), at To endpoint, to endpoint 22387, identity 291->289: 10.0.0.93:2181 -> 10.15.10.151:52378 tcp ACK
-> forward (Established connection), at To endpoint, to endpoint 49574, identity 289->291: 10.15.10.151:52378 -> 10.15.173.191:2181 tcp ACK
-> forward (Established connection), at To endpoint, to endpoint 49574, identity 289->291: 10.15.10.151:52378 -> 10.15.173.191:2181 tcp ACK
-> forward (Reply connection), at To endpoint, to endpoint 22387, identity 291->289: 10.0.0.93:2181 -> 10.15.10.151:52378 tcp ACK
-> forward (Established connection), at To endpoint, to endpoint 49574, identity 289->291: 10.15.10.151:52378 -> 10.15.173.191:2181 tcp ACK
-> forward (Reply connection), at To endpoint, to endpoint 22387, identity 291->289: 10.0.0.93:2181 -> 10.15.10.151:52378 tcp ACK
-> forward (Reply connection), at To host, to endpoint 0, identity 289->1: 10.15.10.151:9092 -> 10.15.28.238:36806 tcp ACK
-> forward (Reply connection), at To endpoint, to endpoint 45959, identity 289->292: 10.0.0.134:9092 -> 10.15.161.37:36344 tcp ACK











^C
Received an interrupt, disconnecting from monitor...

root@minikube:~# cilium monitor
Listening for events on 2 CPUs with 64x4096 of shared memory
Press Ctrl-C to quit



^C
Received an interrupt, disconnecting from monitor...

````

I'm attaching the cilium bugtool from this system.   
[cilium-bugtool-20180125-020501.737+0000-UTC-273827577.tar.zip](https://github.com/cilium/cilium/files/1662339/cilium-bugtool-20180125-020501.737.0000-UTC-273827577.tar.zip)

**Comment by danwent** [comment]
gops output

````
root@minikube:~# gops
32564 31924 gops                 go1.9.2 /bin/gops
252   1     cilium-node-monitor* go1.9   /usr/bin/cilium-node-monitor
1     0     cilium-agent       * go1.9   /usr/bin/cilium-agent
331   1     cilium-health      * go1.9   /usr/bin/cilium-health
282   257   cilium-health      * go1.9   /usr/bin/cilium-health
root@minikube:~# gops stack 252
goroutine 18 [running]:
runtime/pprof.writeGoroutineStacks(0xd68200, 0xc420270010, 0x7efcc996c000, 0x0)
	/usr/local/go/src/runtime/pprof/pprof.go:608 +0xa7
runtime/pprof.writeGoroutine(0xd68200, 0xc420270010, 0x2, 0xc4202cef70, 0xc4204c2430)
	/usr/local/go/src/runtime/pprof/pprof.go:597 +0x44
runtime/pprof.(*Profile).WriteTo(0xd5dbe0, 0xd68200, 0xc420270010, 0x2, 0x0, 0x0)
	/usr/local/go/src/runtime/pprof/pprof.go:310 +0x3ab
github.com/cilium/cilium/vendor/github.com/google/gops/agent.handle(0xd68200, 0xc420270010, 0xc420426000, 0x1, 0x1, 0x0, 0x0)
	/go/src/github.com/cilium/cilium/vendor/github.com/google/gops/agent/agent.go:169 +0x1a3
github.com/cilium/cilium/vendor/github.com/google/gops/agent.listen()
	/go/src/github.com/cilium/cilium/vendor/github.com/google/gops/agent/agent.go:117 +0x2c5
created by github.com/cilium/cilium/vendor/github.com/google/gops/agent.Listen
	/go/src/github.com/cilium/cilium/vendor/github.com/google/gops/agent/agent.go:98 +0x428

goroutine 1 [IO wait, 28 minutes]:
internal/poll.runtime_pollWait(0x7efcc9910df0, 0x77, 0x0)
	/usr/local/go/src/runtime/netpoll.go:173 +0x57
internal/poll.(*pollDesc).wait(0xc4202bc318, 0x77, 0xffffffffffffff00, 0xd69c40, 0xd65fd8)
	/usr/local/go/src/internal/poll/fd_poll_runtime.go:85 +0xae
internal/poll.(*pollDesc).waitWrite(0xc4202bc318, 0xc420115700, 0xa7, 0xee)
	/usr/local/go/src/internal/poll/fd_poll_runtime.go:94 +0x3d
internal/poll.(*FD).Write(0xc4202bc300, 0xc420115770, 0xa7, 0xee, 0x0, 0x0, 0x0)
	/usr/local/go/src/internal/poll/fd_unix.go:226 +0x244
net.(*netFD).Write(0xc4202bc300, 0xc420115770, 0xa7, 0xee, 0x20, 0x0, 0x0)
	/usr/local/go/src/net/fd_unix.go:220 +0x52
net.(*conn).Write(0xc420270000, 0xc420115770, 0xa7, 0xee, 0x0, 0x0, 0x0)
	/usr/local/go/src/net/net.go:188 +0x6d
main.(*Monitor).send(0xddf110, 0xc420058e70, 0x64, 0x64, 0x1, 0x0, 0x9)
	/go/src/github.com/cilium/cilium/monitor/monitor.go:140 +0x420
main.(*Monitor).receiveEvent(0xddf110, 0x7efcc985b1e8, 0x1)
	/go/src/github.com/cilium/cilium/monitor/monitor.go:151 +0xb9
main.(*Monitor).(main.receiveEvent)-fm(0x7efcc985b1e8, 0x1)
	/go/src/github.com/cilium/cilium/monitor/monitor.go:66 +0x3e
github.com/cilium/cilium/pkg/bpf.(*PerfEvent).Read(0xc420094580, 0xc4200c9bf8, 0xc4200c9be8)
	/go/src/github.com/cilium/cilium/pkg/bpf/perf.go:330 +0x1a6
github.com/cilium/cilium/pkg/bpf.(*PerCpuEvents).ReadAll(0xc420446000, 0xc4200c9bf8, 0xc4200c9be8, 0x0, 0x0)
	/go/src/github.com/cilium/cilium/pkg/bpf/perf.go:493 +0x9f
main.(*Monitor).Run(0xddf110, 0x40)
	/go/src/github.com/cilium/cilium/monitor/monitor.go:66 +0x24b
main.runNodeMonitor()
	/go/src/github.com/cilium/cilium/monitor/main.go:87 +0x20d
main.glob..func1(0xd63a20, 0xddf110, 0x0, 0x0)
	/go/src/github.com/cilium/cilium/monitor/main.go:41 +0x20
github.com/cilium/cilium/vendor/github.com/spf13/cobra.(*Command).execute(0xd63a20, 0xc4200101a0, 0x0, 0x0, 0xd63a20, 0xc4200101a0)
	/go/src/github.com/cilium/cilium/vendor/github.com/spf13/cobra/command.go:648 +0x234
github.com/cilium/cilium/vendor/github.com/spf13/cobra.(*Command).ExecuteC(0xd63a20, 0xc420000260, 0x0, 0x0)
	/go/src/github.com/cilium/cilium/vendor/github.com/spf13/cobra/command.go:735 +0x2fe
github.com/cilium/cilium/vendor/github.com/spf13/cobra.(*Command).Execute(0xd63a20, 0x43728c, 0xc420000260)
	/go/src/github.com/cilium/cilium/vendor/github.com/spf13/cobra/command.go:693 +0x2b
main.execute()
	/go/src/github.com/cilium/cilium/monitor/main.go:52 +0x2d
main.main()
	/go/src/github.com/cilium/cilium/monitor/main.go:59 +0x20

goroutine 5 [syscall, 162 minutes]:
os/signal.signal_recv(0x0)
	/usr/local/go/src/runtime/sigqueue.go:131 +0xa6
os/signal.loop()
	/usr/local/go/src/os/signal/signal_unix.go:22 +0x22
created by os/signal.init.0
	/usr/local/go/src/os/signal/signal_unix.go:28 +0x41

goroutine 19 [semacquire, 18 minutes]:
sync.runtime_SemacquireMutex(0xddef74, 0x0)
	/usr/local/go/src/runtime/sema.go:71 +0x3d
sync.(*Mutex).Lock(0xddef70)
	/usr/local/go/src/sync/mutex.go:134 +0xee
main.(*Monitor).handleConnection(0xddf110, 0xd6d480, 0xc420498090)
	/go/src/github.com/cilium/cilium/monitor/monitor.go:103 +0xe1
created by main.runNodeMonitor
	/go/src/github.com/cilium/cilium/monitor/main.go:85 +0x1f3
`````

**Comment by tgraf** [comment]
The `Write()` is blocking and waiting using `poll()` while holding the mutex which is required to accept new connections.

**Comment by tgraf** [comment]
Fixed via #2933
